### PR TITLE
SW-1352 add support for pep 440 in the mr beam plugin

### PR DESCRIPTION
**Pull request by khaledsherkawi**
SW-1352 Refactor the PEP version comparison class

SW-1352 Add frontend and backend methods to handle PEP440 comparisons via API calls

SW-1352 Fix the naming of the Mr Beam Plugin version constants and use the PEP440 comparison method

### CODE REVIEW

File: octoprint_mrbeam/version_comparator.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+from octoprint_mrbeam.mrb_logger import mrb_logger
```
Comment (JFSolution) [inline octoprint_mrbeam/version_comparator.py]: please move the file to the util

File: octoprint_mrbeam/__init__.py [hunk 2]
```diff
@@ -159,6 +161,14 @@ class MrBeamPlugin(
     TIME_NTP_SYNC_CHECK_INTERVAL_FAST = 10.0
     TIME_NTP_SYNC_CHECK_INTERVAL_SLOW = 120.0
 
+    COMPARISON_OPTIONS = [
```
Comment (JFSolution) [inline octoprint_mrbeam/__init__.py]: please move this to the verion_comperator file

File: tests/softwareupdate/test_comparison.py [hunk 3]
```diff
@@ -1,20 +1,19 @@
...
-    VersionComperator,
     _generate_config_of_beamos,
     get_config_for_version,
 )
+from octoprint_mrbeam import MrBeamPlugin
 
 
-def bla(comp1, comparision_options):
-    return VersionComperator.get_comperator(comp1, comparision_options).priority
+def bla(comp1, comparison_options):
```
Comment (JFSolution) [inline tests/softwareupdate/test_comparison.py]: please rename or try to use it without this function

File: octoprint_mrbeam/__init__.py [hunk 4]
```diff
@@ -1306,6 +1307,21 @@ def generate_backlash_compenation_pattern_gcode(self, data):
             res = dict(calibration_pattern=destFile, target=FileDestinations.LOCAL)
             return jsonify(res)
 
+    # simpleApiCommand: compare_pep440_versions;
+    def compare_pep440_versions(self, data):
```
Comment (JFSolution) [inline octoprint_mrbeam/__init__.py]: Can you please add a Google Doc string

File: octoprint_mrbeam/__init__.py [hunk 5]
```diff
@@ -104,6 +106,7 @@
 from octoprint_mrbeam.util.uptime import get_uptime, get_uptime_human_readable
 from octoprint_mrbeam.util import get_thread
 from octoprint_mrbeam import camera
+from .version_comparator import VersionComparator
```
Comment (JFSolution) [inline octoprint_mrbeam/__init__.py]: please change it to octoprint_mrbeam.version.comperator

**Review by JFSolution — CHANGES_REQUESTED** [review]

File: octoprint_mrbeam/version_comparator.py (hunk 1 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/version_comparator.py]: done

File: octoprint_mrbeam/__init__.py (hunk 2 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/__init__.py]: done

File: tests/softwareupdate/test_comparison.py (hunk 3 above)
Comment (khaledsherkawi) [reply tests/softwareupdate/test_comparison.py]: done

File: octoprint_mrbeam/__init__.py (hunk 5 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/__init__.py]: done

File: octoprint_mrbeam/__init__.py (hunk 4 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/__init__.py]: done

File: octoprint_mrbeam/util/version_comparator.py [hunk 6]
```diff
@@ -34,3 +34,12 @@ def get_comparator(comparision_string, comparision_options):
         for item in comparision_options:
             if item.identifier == comparision_string:
                 return item
+
+
+COMPARISON_OPTIONS = [
```
Comment (JFSolution) [inline octoprint_mrbeam/util/version_comparator.py:67]: I would place this at the beginning of the file

File: octoprint_mrbeam/software_update_information.py [hunk 7]
```diff
@@ -515,12 +516,12 @@ def _generate_config_of_beamos(plugin, moduleconfig, beamos_version, tierversion
...
         config_for_beamos_versions.items(),
         key=lambda com: VersionComparator.get_comparator(
-            com[0], plugin.COMPARISON_OPTIONS
+            com[0], COMPARISON_OPTIONS
         ).priority,
     )
 
     config_for_beamos = get_config_for_version(
-        beamos_version, sorted_config_for_beamos_versions, plugin.COMPARISON_OPTIONS
+                                                                                                                                                                                                                    beamos_version, sorted_config_for_beamos_versions, COMPARISON_OPTIONS
```
Comment (JFSolution) [inline octoprint_mrbeam/software_update_information.py]: there is something messy with this line, please check

**Review by JFSolution — CHANGES_REQUESTED** [review]

File: octoprint_mrbeam/software_update_information.py (hunk 7 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/software_update_information.py]: done

File: octoprint_mrbeam/util/version_comparator.py (hunk 6 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/util/version_comparator.py:67]: This shows an error when placed at the top. The IDE marks the class 'VersionComparator' as not defined yet

Comment (JFSolution) [reply octoprint_mrbeam/util/version_comparator.py:67]: okay

**Review by JFSolution — APPROVED** [review]

**Review by JFSolution — COMMENTED** [review]
Did you change anything in prettier? There is a lot of reformating

File: octoprint_mrbeam/__init__.py [hunk 8]
```diff
@@ -1296,6 +1300,33 @@ def generate_backlash_compenation_pattern_gcode(self, data):
...
+        self._logger.info("compare_pep440_versions() request: %s", data)
+        try:
+            v1 = pkg_resources.parse_version(data.get("v1", None))
+            v2 = pkg_resources.parse_version(data.get("v2", None))
+            comparator = data.get("operator", None)
+            result = VersionComparator.get_comparator(
+                comparator, COMPARISON_OPTIONS
+            ).compare(v1, v2)
+            return make_response(json.dumps(result), 200)
+        except Exception as e:
```
Comment (JFSolution) [inline octoprint_mrbeam/__init__.py]: Please handle the exception or catch before if one of the parameters is missing

**Comment by khaledsherkawi** [comment]
> Did you change anything in prettier? There is a lot of reformating

Done - fixed

File: octoprint_mrbeam/__init__.py (hunk 8 above)
Comment (khaledsherkawi) [reply octoprint_mrbeam/__init__.py]: Done

**Review by JFSolution — APPROVED** [review]

**Pull request merged**